### PR TITLE
[Google Blockly] v10 bump!

### PR DESCRIPTION
**Pull request by mikeharv**
This bump our version of Google Blockly to v10.3.1, which is the most recent version!

Major release notes: https://github.com/google/blockly/releases/tag/blockly-v10.0.0

#### Some thing were renamed:
Simplest to fix were some renamed variables, such as `arrow_` to `arrow`.
See commits https://github.com/code-dot-org/code-dot-org/pull/55770/commits/bf5d0fd98b3e9b1fbf8ed4547c37b2ef34494396, https://github.com/code-dot-org/code-dot-org/pull/55770/commits/f7a08d4db0e5d53f2422aab4b105563c42018daa, https://github.com/code-dot-org/code-dot-org/pull/55770/commits/7c440820d00e74ca5f1676a023ac34e2afcca894, https://github.com/code-dot-org/code-dot-org/pull/55770/commits/73555d1fadf15d3a3acb19222cb4797e9ed4a35e for details.

#### New API for Generators
Code generators were refactored here: https://github.com/google/blockly/pull/7150
Where previously, we might have said `generator.sprite_parameter_get`, for a generator function, now it would be `generator.forBlock.sprite_parameter_get`. In a couple of places we can just patch this. Additionally, we can also assign all the properties of our JS generator to `forBlock`. Even though this isn't technically a breaking change, it was causing a flood of deprecation warnings, so we needed to address it. See first commit: https://github.com/code-dot-org/code-dot-org/pull/55770/commits/51af9559ca6799d37b062e782036875e85c2464e

#### Updated plugins
All Blockly plugins were updated to v10 compatible versions. These were tested manually. An exception exists for `@blockly/renderer-inline-row-separators` which was deprecated. See below.

#### Inline Row Separators -> End Row Inputs
A plugin that we have been relying on (for minitoolbox blocks) was deprecated with v10.2.
https://www.npmjs.com/package/@blockly/renderer-inline-row-separators
This changes our logic slightly. The plugin allowed us to arrange multiple inline value input connectors on separate rows by inserting dummy inputs between them (ie. we could put the dummy input with a flyout field below the other inputs, as long as the row also ended with a dummy input).  Now the flyout field's input needs to follow an end row input (https://developers.google.com/blockly/guides/create-custom-blocks/define-blocks#block_inputs). 
This means we need to convert the last input of a block from dummy to end row type. The only difference is that end row inputs always acts like a newline. This gets complicated as these custom blocks are made by our infamous `block_utils` flow. As far as I can tell we _only_ use a dummy at the end of a block so it should be safe to just always use end row inputs now - for all blocks. Again, the only difference is that they will always mark the end of a row.
CDO Blockly has no concept of end row inputs so I added a monkey patch to its wrapper. I opted for this change over the opposite because once Sprite Lab is migrated, the only CDO Blockly still using `block_utils` will be Minecraft.

#### Edit buttons can show in modal function editor
https://codedotorg.atlassian.net/browse/CT-124
The shareable procedures plugin has been updated so that we're now able to edit a function in the modal editor even when it's already open. This work was drafted https://github.com/code-dot-org/code-dot-org/pull/55155 back in November. I included it here because it's the single biggest feature that the v10 bump was blocking. **Note that the buttons to create Functions/Behaviors are hidden from the modal editor's toolbox, but they would now work correctly if added.**
image

#### Flyout fields cleaned up
The v10 update also includes a new render management system. Other than that it should improve rendering efficiency, we shouldn't have to interact with it much. However, Beka and I discovered that it had a couple of bugs that were causing issues when our mini-toolbox blocks would dynamically create their flyout fields. This was be patched in Blockly version `10.3.1`, which is why we are using that version in this branch.
While pairing with Beka, we noticed that some of the code, originally sourced from outside our repo, seemed to be redundant or unnecessary for our implementation. Specifically, we had some leftover logic about showing and hide flyouts. This logic isn't needed because we were never hiding the flyouts; rather we just remove the field and flyout entirely when the user clicks the toggle. A few lines that were easy to remove have been deleted here.

## Links

https://codedotorg.atlassian.net/browse/CT-1
https://codedotorg.atlassian.net/browse/CT-124
https://codedotorg.atlassian.net/browse/CT-130

## Testing story

Some tests were failing initially. Our block utils test had to be updated to use `appendEndRowInput`. We also had some UI tests failures due to changes in how block ids were created. 
Blockly had made a change here: https://github.com/google/blockly/issues/7432
They've ended up reverting this for 10.3.1 for us which is another reason we need to use this specific version.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/addons/cdoFieldImage.js [hunk 1]
```diff
@@ -21,9 +21,9 @@ export default class CdoFieldImage extends GoogleBlockly.FieldImage {
       // the new width and height.
       // TODO: This will need to be updated when upgrading to the 10.0 version of blockly.
```
Comment (molly-moen) [inline apps/src/blockly/addons/cdoFieldImage.js]: remove todo?

File: apps/src/blockly/addons/cdoFieldFlyout.js [hunk 2]
```diff
@@ -93,8 +74,8 @@ export default class CdoFieldFlyout extends GoogleBlockly.Field {
     // On first render, if the flyout is not visible and it should be,
```
Comment (molly-moen) [inline apps/src/blockly/addons/cdoFieldFlyout.js]: update this comment to remove "it should be" since we removed that check?

File: apps/src/blockly/addons/functionEditor.js [hunk 3]
```diff
@@ -101,10 +101,10 @@ export default class FunctionEditor {
     // we have to pass the main ws so that the correct procedures are populated
     // false to not show the new function button inside the modal editor
     this.editorWorkspace.registerToolboxCategoryCallback('PROCEDURE', () =>
-      functionsFlyoutCategory(Blockly.mainBlockSpace, true)
```
Comment (molly-moen) [inline apps/src/blockly/addons/functionEditor.js:104]: why was this changed?

File: apps/src/blockly/customBlocks/googleBlockly/mutators/functionMutatorHelpers.js [hunk 4]
```diff
@@ -7,10 +7,13 @@ import msg from '@cdo/locale';
...
-      .appendDummyInput('DESCRIPTION_ROW')
+      .appendEndRowInput('DESCRIPTION_ROW')
       .appendField(`${msg.description()} `, 'DESCRIPTION_LABEL')
-      .appendField(new Blockly.FieldTextInput(description), 'DESCRIPTION');
-    block.moveInputBefore('DESCRIPTION_ROW', 'STACK');
+      .appendField(new Blockly.FieldMultilineInput(description), 'DESCRIPTION');
+    const inputToPrecede = block.getInput('flyout_input')
+      ? 'flyout_input'
+      : 'STACK';
+    block.moveInputBefore('DESCRIPTION_ROW', inputToPrecede);
```
Comment (molly-moen) [inline apps/src/blockly/customBlocks/googleBlockly/mutators/functionMutatorHelpers.js:16]: why did this need to change?

**Review by molly-moen — COMMENTED** [review]
Looks good, some minor comments/questions

File: apps/src/blockly/addons/cdoRendererThrasos.js [hunk 5]
```diff
@@ -1,9 +1,9 @@
 import GoogleBlockly from 'blockly/core';
 import CdoPathObject from './cdoPathObjectThrasos';
 import CdoConstantsProvider from './cdoConstantsProvider';
-import {addInlineRowSeparators} from '@blockly/renderer-inline-row-separators';
 
-export class CdoRendererThrasosBase extends GoogleBlockly.thrasos.Renderer {
+export default class CdoRendererThrasosBase extends GoogleBlockly.thrasos
+  .Renderer {
```
Comment (ebeastlake) [inline apps/src/blockly/addons/cdoRendererThrasos.js:6]: Is this on a newline intentionally? Or is it just rendering weird in Github?

File: apps/src/blockly/addons/functionEditor.js (hunk 3 above)
Comment (ebeastlake) [reply apps/src/blockly/addons/functionEditor.js:104]: +1 It looks like passing the main workspace here vs. the editor workspace was intentional: `we have to pass the main ws so that the correct procedures are populated`. But we did move away from the `mainBlockSpace` language?

File: apps/src/blockly/customBlocks/googleBlockly/behaviorBlocks.js [hunk 6]
```diff
@@ -184,26 +184,13 @@ GoogleBlockly.Extensions.registerMutator(
  * @param {WorkspaceSvg} workspace The workspace containing procedures.
  * @returns an array of XML block elements
  */
-export function flyoutCategory(workspace, functionEditorOpen = false) {
+export function flyoutCategory(workspace) {
   const blockList = [];
 
-  const behaviorDefinitionBlock = {
```
Comment (ebeastlake) [inline apps/src/blockly/customBlocks/googleBlockly/behaviorBlocks.js:190]: Can you clarify why this was removed? I see it's not currently used, but I don't remember when we stopped needing it.

File: apps/src/blockly/customBlocks/googleBlockly/proceduresBlocks.js [hunk 7]
```diff
@@ -112,7 +112,6 @@ GoogleBlockly.Extensions.register('procedures_edit_button', function () {
     Blockly.useModalFunctionEditor &&
     this.inputList.length &&
     !this.workspace.isFlyout &&
-    this.workspace.id !== Blockly.functionEditor.getWorkspaceId() &&
     toolboxConfigurationSupportsEditButton(this)
```
Comment (ebeastlake) [inline apps/src/blockly/customBlocks/googleBlockly/proceduresBlocks.js:113]: I can't comment on the actual line (109), but it looks like there's a TODO here that could use clean-up?

File: apps/src/blockly/customBlocks/googleBlockly/proceduresBlocks.js [hunk 8]
```diff
@@ -233,30 +232,24 @@ GoogleBlockly.Extensions.registerMixin(
...
-    const newFunctionButton = getNewFunctionButtonWithCallback(
-      workspace,
-      functionDefinitionBlock
-    );
+  if (Blockly.useModalFunctionEditor) {
+    const newFunctionButton = getNewFunctionButtonWithCallback(workspace);
     blockList.push(newFunctionButton);
   } else {
+    // Note: Blockly.Msg was undefined when this code was extracted into global scope
+    const functionDefinitionBlock = {
```
Comment (ebeastlake) [inline apps/src/blockly/customBlocks/googleBlockly/proceduresBlocks.js]: nit that doesn't need to be addressed: Since this is a block "constant"/"template" that I would have extracted into a `constants` file if I could have, it could also have continued to live above where it's used, but this also makes sense and no need to revert!

File: apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js [hunk 9]
```diff
@@ -55,7 +53,6 @@ export const blocks = {
       if (!block.getInput(INPUTS.FLYOUT)) {
         const flyoutField = createFlyoutField(block);
         flyoutField.showEditor();
-        flyoutField.render_();
```
Comment (ebeastlake) [inline apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js:58]: Is this part of the flyout fields clean-up?

File: apps/src/blockly/addons/functionEditor.js (hunk 3 above)
Comment (mikeharv) [reply apps/src/blockly/addons/functionEditor.js:104]: We need to register this category using the editor workspace so that when we `getNewFunctionButtonWithCallback`, it register a button callback for the editor workspace. Otherwise, the button doesn't do anything if it's in the flyout for the modal editor. Note that we also register these category callbacks on the main workspace in the wrapper. We no longer have to look at the main workspace to find the procedures, so I'll removed that comment the Emily is quoting above.

File: apps/src/blockly/customBlocks/googleBlockly/mutators/functionMutatorHelpers.js (hunk 4 above)
Comment (mikeharv) [reply apps/src/blockly/customBlocks/googleBlockly/mutators/functionMutatorHelpers.js:16]: Good catch; apologies for missing this in the PR write up! I did this to better align to CDO Blockly's modal editor. ie. the flyout is below the description field (and closer to the definition blocks)
image

I brought this up on Slack on 12/14: https://codedotorg.slack.com/archives/C05HZFH7BED/p1702585446512979

File: apps/src/blockly/addons/cdoRendererThrasos.js (hunk 5 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoRendererThrasos.js:6]: I think it's just due to Prettier determining that the line is now a little too long. Pretty sure that all I did was delete `default` and press save.

Confirmed by trying to manually revert:
image

File: apps/src/blockly/customBlocks/googleBlockly/proceduresBlocks.js (hunk 7 above)
Comment (mikeharv) [reply apps/src/blockly/customBlocks/googleBlockly/proceduresBlocks.js:113]: Good catch, thanks!

File: apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js (hunk 9 above)
Comment (mikeharv) [reply apps/src/blockly/customBlocks/googleBlockly/spritelabBlocks.js:58]: Yes! It was actually just redundant. Normally `render_` should only be called by `getSize`.

File: apps/src/blockly/addons/functionEditor.js (hunk 3 above)
Comment (ebeastlake) [reply apps/src/blockly/addons/functionEditor.js:104]: So we don't care if the "new function"/"new behavior" button is rendered inside the editor toolbox? I see it in your screenshots and I don't think that's how it worked before (and I imagine, if we could avoid that, we'd want to?).

Comment (ebeastlake) [reply apps/src/blockly/addons/functionEditor.js:104]: Screenshot 2024-01-17 at 4 25 37 PM
Screenshot 2024-01-17 at 4 25 58 PM

```
We need to register this category using the editor workspace so that when we getNewFunctionButtonWithCallback, it register a button callback for the editor workspace. Otherwise, the button doesn't do anything if it's in the flyout for the modal editor.
``` 
But you're saying we want this button? I thought we a) didn't want it and b) didn't want it to do anything. (Sorry for two comments, my 4:30 PM brain is dragging.)

Comment (mikeharv) [reply apps/src/blockly/addons/functionEditor.js:104]: Do you recall a reason why we wouldn't to be able to create a function/behavior from within the editor? I don't think there's a technical reason we can't, so maybe it's worth checking with Amy?

On CDO Blockly, you are right that you can't create a new function if the function editor is open (same for behaviors). Strangely, you _can_ create a function if just the behavior editor is open (and vice versa). If you try this, things get gross, because CDO makes two modals which it isn't really meant to. :( 

I'm not saying I know for sure, but I wonder if the decision to hide these buttons in the original modal editor was a technical one or a design decision.

File: apps/src/blockly/customBlocks/googleBlockly/behaviorBlocks.js (hunk 6 above)
Comment (mikeharv) [reply apps/src/blockly/customBlocks/googleBlockly/behaviorBlocks.js:190]: It was used to define an empty c-shaped behavior definition block that was placed in the toolbox when the modal editor was disabled. We do this for functions, but not behaviors because behaviors are always hidden from the main workspace. I originally added this back in the summer as a way of testing our earliest behavior blocks.

File: apps/src/blockly/addons/functionEditor.js (hunk 3 above)
Comment (molly-moen) [reply apps/src/blockly/addons/functionEditor.js:104]: > We no longer have to look at the main workspace to find the procedures

Why is this?

Comment (ebeastlake) [reply apps/src/blockly/addons/functionEditor.js:104]: @mikeharv I think it was a design decision, but I'm not sure it's one I disagree with?

Comment (mikeharv) [reply apps/src/blockly/addons/functionEditor.js:104]: @molly-moen More detailed answer in Slack, but we always populate the flyout by looking for procedure blocks on both the main and hidden workspaces. I think the comment has long been out of date.

@ebeastlake I've made a partial revert so that we again only add the "Create" buttons to the flyout if we're on the main workspace. I've left in the change that would make these buttons work correctly if we did want to use them in the modal editor's toolbox at any point.

File: dashboard/test/ui/features/code_tools/google_blockly/modal_function_editor.feature [hunk 10]
```diff
@@ -31,7 +31,7 @@ Scenario: Can edit a function
   # Open Sprites flyout
   And I press "blockly-d"
   # Drag block to top of function
-  And I drag block "new-sprite-block" to offset "40, 100"
+  And I drag block "new-sprite-block" to offset "40, 105"
```
Comment (molly-moen) [inline dashboard/test/ui/features/code_tools/google_blockly/modal_function_editor.feature]: I think you'll need to make the same update [here](https://github.com/code-dot-org/code-dot-org/blob/3c512ad40adb5e38a02e9918254efa3c5358af1b/dashboard/test/ui/features/code_tools/google_blockly/modal_function_editor_eyes.feature#L21). It would be a good idea to run that test and [this one](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/test/ui/features/code_tools/google_blockly/spritelab_eyes.feature) locally to check that the eyes tests are still working as expected.

Comment (mikeharv) [reply dashboard/test/ui/features/code_tools/google_blockly/modal_function_editor.feature]: Oh good call out. Do you think I didn't run into this yet because drone doesn't run eyes tests?

Comment (molly-moen) [reply dashboard/test/ui/features/code_tools/google_blockly/modal_function_editor.feature]: yep!

**Review by molly-moen — APPROVED** [review]
yay v10! 🎉

File: dashboard/test/ui/features/code_tools/google_blockly/modal_function_editor.feature (hunk 10 above)
Comment (mikeharv) [reply dashboard/test/ui/features/code_tools/google_blockly/modal_function_editor.feature]: I initially changed the other test as you suggested, but then reverted both as we decided not to use the multi-line field until this is resolved: https://github.com/google/blockly/issues/7209
